Fix: Static fonts configured in 4.19.1 continue working after upgrade

mpconley opened the pull request:
#### Brief overview of PR changes/additions

Users who configured static fonts like "EB Garamond SemiBold" in Mudlet 4.19.1 will have their fonts continue working after upgrading to 4.20.

#### Motivation for adding to Mudlet

Ensures a smooth upgrade experience for players using static font variants.

#### Other info (issues closed, discussion etc)

Fixes #8248

mpconley committed:
Fix: Font antialiasing and backward compatibility with static fonts

Fixes #8248

- Add Host::createFontWithSettings() to preserve antialiasing when creating fonts
- Add Host::parseFontNameAndStyle() for backward compatibility with static fonts
- Update setFont() to automatically fallback to base font family with weight
- Ensures smooth upgrade for users with static font configurations
- Maintains antialiasing preferences across all font operations

mpconley committed:
Merge branch 'development' into fix-font-antialiasing-static-fonts

mpconley committed:
Merge branch 'development' into fix-font-antialiasing-static-fonts

mpconley committed:
Improve font rendering at small sizes with size-dependent hinting

- Add hinting preferences to address artifacts at font sizes < 12pt
- Small fonts use default hinting for better clarity
- Large fonts use minimal hinting for smoother antialiasing
- Update TFontAttributes to support hinting preferences

mpconley committed:
Windows: Use PreferVerticalHinting for better ClearType antialiasing

Windows 10 users were still experiencing antialiasing/scaling issues
despite the previous font rendering improvements. This is because
Windows uses ClearType subpixel antialiasing which works best with
vertical hinting rather than the default/no hinting used on other
platforms.

Changes:
- Host::createFontWithSettings() now uses PreferVerticalHinting on
  Windows for antialiased fonts instead of size-dependent hinting
- TFontAttributes constructor and setAntiAliasOption() updated to
  use PreferVerticalHinting on Windows
- TFontAttributes::makeFont() uses platform-specific hinting:
  * Windows: Always PreferVerticalHinting for antialiased fonts
  * Other platforms: Size-dependent (Default <12pt, None >=12pt)

This leverages Windows' ClearType technology for optimal subpixel
rendering while maintaining the existing behavior on macOS and Linux.

mpconley committed:
Merge branch 'development' into fix-font-antialiasing-static-fonts

mpconley committed:
Add setLabelFont() API and refactor font creation

- Refactor Host::createFontWithSettings() to delegate to TFontAttributes::createFont()
- Centralize platform-specific font hinting logic in TFontAttributes::makeFont()
- Add TLabel::setFontFromProfile() to apply profile's antialiasing settings to labels
- Add setLabelFont() Lua function for applying profile antialiasing to labels
- Update Geyser.Label:setFont() and setFontSize() to automatically use profile antialiasing
- Fixes Windows 10 label antialiasing/scaling issues reported in Discord
- Backwards compatible: existing code continues to work, new function is opt-in

mpconley committed:
Refactor applyProfileAntialiasing() to use helper method

- Add TLabel::updateFont() helper that checks mUseProfileAntialiasing flag
- Simplify applyProfileAntialiasing() to just set flag and call updateFont()
- When disabling, properly restore default Qt font rendering (not profile AA)
- Add boolean parameter to applyProfileAntialiasing() Lua function
- Update Geyser.Label to pass enable parameter correctly
- Architecture: Flag-based approach with centralized font update logic

mpconley committed:
Fix: Apply size-dependent hinting on Windows for small fonts

- Small fonts (< 12pt) now use PreferDefaultHinting on Windows
- Larger fonts (>= 12pt) continue to use PreferVerticalHinting for ClearType
- Addresses Discord feedback about scaling/antialiasing issues at small font sizes
- Fixes rendering quality for serif fonts at size 10-12pt on Windows

mpconley committed:
Merge branch 'development' into fix-font-antialiasing-static-fonts

mpconley committed:
Fix: Apply antialiasing in Geyser.Label constructor

When useProfileAntialiasing=true is set in the constructor options,
the font antialiasing settings weren't being applied because the
constructor only set the font property without calling setFont().

This fix checks if both useProfileAntialiasing and font are set
during construction and calls setFont() to trigger the antialiasing
application, ensuring Windows users see properly antialiased text
when using custom fonts like 'EB Garamond' with useProfileAntialiasing.

mpconley committed:
Merge branch 'development' into fix-font-antialiasing-static-fonts

mpconley committed:
Merge branch 'development' into fix-font-antialiasing-static-fonts

mpconley committed:
Remove unvalidated label antialiasing features

Retain only the core fix for issue #8248 (static font backward compatibility):
- Keep Host::parseFontNameAndStyle() for parsing static font names
- Keep Host::createFontWithSettings() for font creation with antialiasing
- Keep setFont() fallback to base font family with weight

Remove features not validated by Discord users:
- Remove applyProfileAntialiasing() Lua function
- Remove TLabel::applyProfileAntialiasing() and updateFont() methods
- Remove Geyser.Label:setUseProfileAntialiasing() method
- Remove useProfileAntialiasing property from Geyser.Label

mpconley committed:
Refine: Remove hinting preference changes, keep only static font support

Revert TConsole.h hinting preference changes that were not directly
related to issue #8248 (static font backward compatibility).

Update Host::createFontWithSettings() to use TFontAttributes directly
rather than relying on the removed static createFont() method.

This commit focuses the PR on the core fix: parsing font names with
embedded style info (e.g. 'EB Garamond SemiBold') to support users
who configured static fonts in Mudlet 4.19.1.

mpconley committed:
Style: Add blank lines for readability in setFont()

mpconley commented:
For testing, installed [EBGaramond-SemiBold.ttf](https://fonts.google.com/specimen/EB+Garamond) and ran this script.

<img width="948" height="894" alt="Screenshot 2025-11-28 at 12 23 08 PM" src="https://github.com/user-attachments/assets/d0e05bd6-4fb8-4aa5-ab4b-90d1a39bd095" />

mpconley committed:
Merge branch 'development' into fix-font-antialiasing-static-fonts

mpconley commented:
```lua
local fontName = "EB Garamond SemiBold"
local result = setFont(fontName)
if result then
    print("Font set successfully!")
    print("Current font: " .. getFont())
else
    print("Failed to set font")
end
```

mpconley committed:
Merge branch 'development' into fix-font-antialiasing-static-fonts

vadi2 commented:
I think I went with a similar approach in https://github.com/vadi2/Mudlet/commit/777f5d0aff7e16e290df96189e454dae8d038293, but the trouble I ran into is that the font still was not available in a label, so scripts from 4.19 broke in 4.20. Did you have better luck here?

mpconley committed:
Fix: Enable static font compatibility in Geyser.Label

- Add automatic setFont() calls in setFont() and setFontSize() methods
- Ensures Geyser.Labels use profile antialiasing settings for static fonts
- Addresses vadi2's concern about label font compatibility in PR #8545
- Maintains backward compatibility for 4.19.1 scripts using static font names
- Uses unified setFont() API instead of separate label-specific functions

mpconley commented:
@vadi2 Thanks for catching this! You're absolutely right that static fonts weren't working in labels, which would have broken the 4.19.1→4.20 upgrade path.

**Fixed in commit 30278a67**: I've implemented automatic static font compatibility for Geyser.Labels by adding `setFont(self.name, font)` calls to the `setFont()` and `setFontSize()` methods in `GeyserLabel.lua`.

**How it works:**
- When users call `geyserLabel:setFont("EB Garamond SemiBold")`, it now automatically calls the existing `setFont(labelName, fontName)` function
- This leverages the same static font parsing logic (`Host::parseFontNameAndStyle()`) that fixed console fonts
- No separate API needed - uses unified approach through existing `setFont()` function

**Tested and verified:**
- ✅ Console fonts: `setFont("EB Garamond SemiBold")` works (original fix)
- ✅ TLabel fonts: `setFont("labelName", "EB Garamond SemiBold")` works  
- ✅ Geyser.Label fonts: Automatic integration maintains compatibility
- ✅ 4.19.1 scripts with static fonts in labels will continue working

The terminal output shows the parsing working correctly:
```
setFont(): Font "EB Garamond SemiBold" not found, using "EB Garamond" with weight QFont::DemiBold
```

This ensures the upgrade path remains intact while using a clean, unified API approach. 🎯

[quick_font_test.txt](https://github.com/user-attachments/files/23951989/quick_font_test.txt)

vadi2 commented:
Cherry-picked into release-4.20 branch.
